driver-core/src/main/java/com/datastax/driver/core/Connection.java

```diff
@@ -452,14 +456,15 @@ private void notifyOwnerWhenDefunct() {
     }
 
     String keyspace() {
-        return keyspace;
+        SetKeyspaceAttempt currentAttempt = targetKeyspace.get();
+        return currentAttempt == null ? null : currentAttempt.keyspace;
```

Comment:
targetKeyspace gets set to null if the future fails, should it not do that?  The reason it does it is if set keyspace fails, we want to provide another opportunity to set the same keyspace.